### DISCUSSION TITLE
Provide the ability to pass a session as a function in the repository.

### DISCUSSION THREAD

**Discussion by Kavan72** [litestar-org/advanced-alchemy#316] [Ideas]
Currently, there is no way to pass a session object at runtime. Ideally, we should be able to create a repository object and pass a session or inject a session while calling a repository method. We have 45 services and 57 repositories, and this feature could help us access any service from any other service and do the same with repositories.

Before I became familiar with this library, I was implementing it like this:

```py

class Database:
    session: AsyncSession

    def __init__(
        self,
        database: str,
        username: str,
        password: str,
        host: str,
        port: int = 5432,
        echo: bool = True,
    ) -> None:
        self.url = URL.create(
            drivername="postgresql+asyncpg",
            username=username,
            password=password,
            host=host,
            port=port,
            database=database,
        ).render_as_string(hide_password=False)
        self.engine = create_async_engine(
            self.url,
            echo=echo,
            echo_pool="debug",
        )
        self.session_maker: async_sessionmaker[AsyncSession] = async_sessionmaker(
            self.engine, expire_on_commit=False
        )

class DatabaseRepository:
    """Repository for performing database queries."""

    def __init__(self, app, model: type[Model]):
        self.app = app
        self.model = model

    async def execute(self, *args, **kwargs):
        return await self.app.database.session.execute(*args, **kwargs)
        
class UserService:
    _model = User

    def __init__(self, app):
        self.app = app
        self.repository = DatabaseRepository(app, self._model)
        
async def get_db_session(request: Request) -> AsyncGenerator[AsyncSession, None]:
    async with request.app.database.session_maker() as session:
        try:
            request.app.database.session = session
            yield session
            await session.commit()
        except exc.SQLAlchemyError as error:
            logger.debug(f"Rolling back due to exception {error}.")
            await session.rollback()
            raise error
        finally:
            await session.close()
            request.app.database.session = None

@router.delete("/user/{userId}")
async def delete_user_handler(
    request: Request,
    session: AsyncSession = Depends(get_db_session),
    user_id: int = Path(..., alias="userId"),
    token: AccessToken = Depends(JWTAPIAccessTokenBearer()),
):
    if not await request.app.services.user_service.delete_user(user_id):
        raise NotFound("User not found")
    return Response(status_code=204)
```

Loading the Database State

The database is loaded like this:
```py
app.database = Database(
    app.config.POSTGRES_DB,
    app.config.POSTGRES_USER,
    app.config.POSTGRES_PASSWORD,
    app.config.POSTGRES_HOST,
    echo=app.config.ENVIRONMENT != ENVIRONMENT.PRODUCTION,
)
```
Loading Services into the Application State

Services are loaded into the application state during startup using a for loop.

**Comment by Priestch** [comment]
I'm also struggling to use other services in one service, some business logic involved multiple database models and using relationship is not enough.

Currently, I'm trying to abstract a universal layer to place all the business logic there, it has methods to provide all the advanced-alchemy services needed or use the services injected in the router handler. The layer will be initialized with database session, it combines the services to do the business. I'm not sure whether it follows the best practice.

If some business logic involved with multiple database models, what's the best practice?

**Reply by cofin** [reply]
@Priestch your best here is to add one layer on top of the services you want to encapsulate.  Here's an example where the logic for using multiple services at at the controller level - but you could imagine this being a new "MutliService" class you create:
```py
    @post(operation_id="AccountRegister", path=urls.ACCOUNT_REGISTER)
    async def signup(
        self,
        request: Request,
        users_service: UserService,
        roles_service: RoleService,
        data: AccountRegister,
    ) -> User:
        """User Signup."""
        user_data = data.to_dict()
        role_obj = await roles_service.get_one_or_none(slug=slugify(users_service.default_role))
        if role_obj is not None:
            user_data.update({"role_id": role_obj.id})
        user = await users_service.create(user_data)
        request.app.emit(event_id="user_created", user_id=user.id)
        return users_service.to_schema(user, schema_type=User)
```

The idea is that you share a single DB session between them.  If you aren't sharing a DB session between the services, you may want to consider using the `auto_purge` as well.  The objects will be re-attached to whatever session you are using next time they are passing into a service.

**Comment by Alc-Alc** [comment]
> Ideally, we should be able to create a repository object and pass a session

Maybe I am misunderstanding, but you can already do this, the repositories and services both accept a session during creation

https://github.com/litestar-org/advanced-alchemy/blob/main/advanced_alchemy/repository/_async.py#L82

https://github.com/litestar-org/advanced-alchemy/blob/main/advanced_alchemy/service/_sync.py#L117

> inject a session while calling a repository method.

If you mean an interface like `repo.get(db_session)`, currently as far as I know, the repository does not rely on anything in the session state, so this can be considered a feature request but I am not sure. @cofin wdyt?

**Reply by cofin** [reply]
@Kavan72 I don't think saving an instance in the state for the duration of the ASGI application is the right approach.  You should let these instances be garbage collected and instantiated as needed using dependency injection.  Also, there's quite a bit of extra stuff you don't need.  For instance, all of your existing code can be replaced by this:
```py
sqlalchemy_config = SQLAlchemyAsyncConfig(
    connection_string="sqlite+aiosqlite:///test.sqlite",
    session_config=AsyncSessionConfig(expire_on_commit=False),
    commit_mode="autocommit",
)
app = FastAPI()
alchemy = AdvancedAlchemy(config=sqlalchemy_config, app=app)
DatabaseSession = Annotated[AsyncSession, Depends(alchemy.provide_session())]
UserService = Annotated[UserService, Depends(provide_users_service)]


async def provide_users_service(db_session: DatabaseSession) -> AsyncGenerator[AuthorService, None]:
    """This provides the default Authors repository."""
    async with UserService.new(session=db_session) as service:
        yield service

@router.delete("/user/{userId}")
async def delete_user_handler(
    user_id: int = Path(..., alias="userId"),
    token: AccessToken = Depends(JWTAPIAccessTokenBearer()),
    user_service: UserService,
):
    if not await service.delete(user_id):
        raise NotFound("User not found")
    return Response(status_code=204)
```

I think you want some sort of management class that helps instantiate these objects or generates closures for you.
